src/qutip_cupy/dense_functions.py

```diff
@@ -31,6 +31,83 @@ def trace_cupydense(cp_arr):
     return cp.trace(cp_arr._cp).item()
 
 
+def _check_shape_inner(left, right):
+    if (left.shape[0] != 1 and left.shape[1] != 1) or right.shape[1] != 1:
+        raise ValueError(
+            "incompatible matrix shapes " + str(left.shape) + " and " + str(right.shape)
+        )
+
+
+def inner_cupydense(left, right, scalar_is_ket=False):
+
+    _check_shape_inner(left, right)
+
+    if left.shape[0] == left.shape[1] == right.shape[1] == 1:
+        if not cp.all(left._cp == 0) and not cp.all(right._cp == 0):
+            print(left._cp)
+            print(right._cp)
+            return (
+                (cp.conj(left._cp[0, 0]) * right._cp[0, 0]).item()
+                if scalar_is_ket
+                else (left._cp[0, 0] * right._cp[0, 0]).item()
+            )
+        return 0.0j
+
+    if left.shape[0] == 1:
+        # TODO check if this runs faster using qutip cublas methods
+        # which allow to pass a flag to take conj
+        return cp.vdot(cp.conj(left._cp), right._cp).item()
```

Comment:
If `left.shape[0] == 1` then `left` is a ket and so we shouldn't call `conj(left)`. Similarly the case below is missing a `conj`. Not sure how tests passed though, so I'm a bit confused.

Reply:
`left` is a bra if `left.shape[0] == 1`, but the rest of the sentence seems to be correct - if it's a bra already, you don't need to take the conjugate.

Reply:
oh wait, it's because `vdot` tries to do the "right thing" when it comes to complex numbers.  In this case, you probably just mean to use `cp.dot` instead of `cp.vdot`, so you don't have to have the duplicated conjugation.

Reply:
@jakelishman Tx for digging into this.
@MrRobot2211 Let's implement Jake's suggestion and avoid the double `conj` by using `cp.dot` here. We can keep `cp.vdot` for the ket case (just maybe add a comment that `vdot` does the conj for us).

Reply:
"digging into" is pretty strong terminology for "I think this is what numpy does, and I assume cupy copies its interface", haha.